pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +131,477 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
-	r.drainer.IgnoreAllDaemonSets = true
+			statusAnnoData.Status = NodeStateLeaseNotOwnedByNode
+			r.setStatusAnnotation(node, statusAnnoData)
 
-	//Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.
-	r.drainer.GracePeriodSeconds = -1
+			err := fmt.Errorf("lease got acquired by someone while request was active. node: %s", nodeName)
+			return true, &reconcile.Result{}, err
+		}
 
-	// TODO - add logical value or attach from the maintancene CR
-	//The length of time to wait before giving up, zero means infinite
-	r.drainer.Timeout = time.Minute
+	case LeaseStatusFail:
+		{
+			r.reqLogger.Errorf("failed to get lease object. node: %s error: %v", nodeName, err)
 
-	// TODO - consider pod selectors (only for VMIs + others ?)
-	//Label selector to filter pods on the node
-	//r.drainer.PodSelector = "kubevirt.io=virt-launcher"
+			statusAnnoData.Status = NodeStateFailedToGetLease
+			r.setStatusAnnotation(node, statusAnnoData)
 
-	cs, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return err
+			return true, &reconcile.Result{}, err
+		}
 	}
-	r.drainer.Client = cs
-	r.drainer.DryRun = false
 
-	return nil
+	// TODO:  how to handle possible failure to modify lease time ?
+	r.adjustLeaseTimeIfRequired(specAnnoData, statusAnnoData, lease)
+
+	return false, nil, nil
 }
 
-var _ reconcile.Reconciler = &ReconcileNodeMaintenance{}
+func (r *ReconcileNodeMaintenance) adjustLeaseTimeIfRequired(specAnnoData *ReconcileNodeMaintenanceSpecInfo, statusAnnoData *ReconcileNodeMaintenanceStatusInfo, lease *coordv1beta1.Lease) error {
+	duration := statusAnnoData.OpDeadline.Sub(statusAnnoData.OpStart)
+	durationSeconds := int64(duration.Nanoseconds()) / int64(time.Second)
 
-type ReconcileHandler interface {
-	StartPodInformer(node *corev1.Node, stop <-chan struct{}) error
+	diffTime := int64(specAnnoData.RequestedMaintenanceWindowSeconds) - durationSeconds
+
+	if diffTime < -3 || diffTime > 3 {
+		r.reqLogger.Infof("Lease time has been changed. %d", int(durationSeconds))
+
+		specSeconds := int32(durationSeconds)
+		lease.Spec.LeaseDurationSeconds = &specSeconds
+
+		if err := r.client.Update(context.TODO(), lease); err != nil {
+			r.reqLogger.Infof("failed to update lease node: %s error: %v", lease.ObjectMeta.Name, err)
+			return err
+		}
+	}
+	return nil
 }
 
-// ReconcileNodeMaintenance reconciles a NodeMaintenance object
-type ReconcileNodeMaintenance struct {
-	// This client, initialized using mgr.Client() above, is a split client
-	// that reads objects from the cache and writes to the apiserver
-	client      client.Client
-	scheme      *runtime.Scheme
-	drainer     *drain.Helper
-	podInformer cache.SharedInformer
+func (r *ReconcileNodeMaintenance) processMaintModeOn(statusAnnoData *ReconcileNodeMaintenanceStatusInfo, node *corev1.Node) (reconcile.Result, error) {
+
+	drainer, _, err := r.initDrainer()
```

Comment:
Is there something about `drainer` that makes it unsuitable to be cached?

Reply:
in file sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go

Reply:
in this project we have MaxConcurrentReconciles = 1 (as configuration didn't set this value), so it should be ok, until someone changes that.

Reply:
fixed